languages/elixir/exercises/concept/regular-expressions/.docs/introduction.md

```diff
@@ -0,0 +1,76 @@
+Regular expressions are a powerful tool for string in Elixir. Regular expressions in Elixir follow the **PCRE** specification (**P**erl **C**ompatible **R**egular **E**xpressions). String patterns representing the regular expression's meaning are first compiled then used for matching all or part of a string. The `=~/2` operator is useful to perform a boolean match on a string. A brief overview of regular expressions follows:
+
+To match _literal characters_, you can use the string as the pattern:
+
+```elixir
+regex = "foo-bar" |> Regex.compile!()
```

Comment:
I don't think I would ever write `Regex.compile!()`. We should start with the `r` sigil here and use it everywhere instead of `compile!`. `compile!` doesn't even need to be mentioned at all in this exercise. From the top of my head, I don't remember why I would need it.

Reply:
I don't think I would use it regularly either, but I wanted students to focus more on the pattern than on making a regex.  If we use the sigil in the examples up front, then it might lead them down some rabbit hole when doing the exercise when I just want them to write patterns.  Also It's an unfortunate reality that you have to escape special characters (e.g. "\\d" rather than "\d") in many other languages and even elixir when using the pattern in certain ways (like with the replace).

Reply:
Also I think it is that much harder to test since there isn't a way to get the regex pattern back once compiled to a regex.  Sure, could maybe use the analyzer, but I think that makes testing extremely complicated.

Reply:
> If we use the sigil in the examples up front, then it might lead them down some rabbit hole when doing the exercise when I just want them to write patterns

I don't really see what complications using the sigil might introduce. Can you give any example?

>  Also It's an unfortunate reality that you have to escape special characters (e.g. "\d" rather than "\d") in many other languages and even elixir when using the pattern in certain ways (like with the replace).

Does that have anything to do with `compile!` vs the `r` sigil, or was that more of a general remark?

> Also I think it is that much harder to test 

I'm not suggesting changing the tests nor the expected solution, I'm only suggesting to change the examples in the introduction.

I think one of the goals of teaching Elixir is to teach writing idiomatic Elixir code. The documentation clearly prefers the sigil to write regular expressions (https://hexdocs.pm/elixir/Regex.html#content) and for this reason I strongly feel we should do that too.

Reply:
Agree, the `~r` sigil is dominant throughout the regex documentation where there is an option of using either.

My concern is with looking at digits `~r/\d/` might suggest that when writing the pattern as a string a student might use "\d" and wonder why this isn't working.  If I have the tests _escape_ the string first (probably using Regex.escape) that might make it too magical?

I think there are two options, and I am now leaning more towards your suggestion:
1. (preference) use ~r in all introduction examples, but noting that when writing patterns as strings, you will either have to manually escape them or use Regex.escape before compile.
2 Write the patterns as strings, with escaped characters, but noting that ~r is a common shortcut and do not have to escape them.

I can see what it all looks like with the former option.